### PR TITLE
SONAR-7899 concept of "startup follower"

### PR DESCRIPTION
**Pull request by simonbrandhof**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/platform/UrlSettings.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+import static org.sonar.api.CoreProperties.SERVER_BASE_URL;
+import static org.sonar.server.app.TomcatContexts.PROPERTY_CONTEXT;
+
+@ComputeEngineSide
+@ServerSide
+public class UrlSettings {
+
+  private static final int DEFAULT_PORT = 9000;
+  private static final int DEFAULT_HTTP_PORT = 80;
+  private static final String ALL_IPS_HOST = "0.0.0.0";
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/platform/UrlSettings.java]: CRITICAL Make this IP "0.0.0.0" address configurable. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1313)

File: server/sonar-server/src/main/java/org/sonar/server/platform/web/RailsAppsDeployer.java [hunk 2]
```diff
@@ -87,15 +85,12 @@ static void deployRailsApp(File appsDir, final String pluginKey, ClassLoader app
...
-      ClassLoaderUtils.copyResources(appClassLoader, pathToRubyInitFile(pluginKey), appDir, new Function<String, String>() {
-        @Override
-        public String apply(@Nullable String relativePath) {
-          // Relocate the deployed files :
-          // relativePath format is: org/sonar/ror/sqale/app/controllers/foo_controller.rb
-          // app path is: org/sonar/ror/sqale
-          // -> deployed file is app/controllers/foo_controller.rb
-          return StringUtils.substringAfter(relativePath, pluginKey + "/");
-        }
+      ClassLoaderUtils.copyResources(appClassLoader, pathToRubyInitFile(pluginKey), appDir, relativePath -> {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/platform/web/RailsAppsDeployer.java]: MAJOR Remove useless curly braces around statement and then remove useless return keyword [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1602)

File: server/sonar-server/src/main/java/org/sonar/server/util/ClassLoaderUtils.java [hunk 3]
```diff
@@ -86,7 +81,7 @@ public boolean apply(@Nullable String path) {
    * @param predicate
    * @return a list of relative paths, for example {"org/sonar/sqale", "org/sonar/sqale/foo", "org/sonar/sqale/foo/bar.txt}. Never null.
    */
-  static Collection<String> listResources(ClassLoader classLoader, String rootPath, Predicate<String> predicate) {
+  public static Collection<String> listResources(ClassLoader classLoader, String rootPath, Predicate<String> predicate) {
     String jarPath = null;
     JarFile jar = null;
     try {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/util/ClassLoaderUtils.java]: MAJOR Change this "try" to a try-with-resources. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2093)

File: server/sonar-server/src/test/java/org/sonar/server/platform/ServerImplTest.java [hunk 4]
```diff
@@ -19,299 +19,280 @@
...
-import org.junit.rules.TemporaryFolder;
-import org.sonar.api.CoreProperties;
-import org.sonar.api.config.Settings;
-import org.sonar.process.ProcessProperties;
-
-import static org.assertj.core.api.Assertions.assertThat;
-import static org.junit.Assert.assertThat;
-import static org.sonar.api.CoreProperties.SERVER_BASE_URL;
-
 public class ServerImplTest {
```
Comment (SonarTech) [inline server/sonar-server/src/test/java/org/sonar/server/platform/ServerImplTest.java]: MAJOR Add some tests to this class. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2187)

File: server/sonar-server/src/main/java/org/sonar/server/platform/ServerSettingsImpl.java [hunk 5]
```diff
@@ -35,17 +35,15 @@
  * <li>environment variables</li>
  * <li>system properties</li>
  * </ol>
- *
- * @since 2.12
  */
-public class WebServerSettings extends Settings implements ServerSettings {
+public class ServerSettingsImpl extends Settings implements ServerSettings {
 
   private final Properties properties;
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/platform/ServerSettingsImpl.java]: MAJOR "properties" is the name of a field in "Settings". [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2387)

File: server/sonar-server/src/main/java/org/sonar/server/platform/StartupMetadataProvider.java [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+   */
+  private StartupMetadata load(DbClient dbClient) {
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      String startupId = selectProperty(dbClient, dbSession, CoreProperties.SERVER_ID);
+      String startedAt = selectProperty(dbClient, dbSession, PROPERTY_STARTED_AT);
+      return new StartupMetadata(startupId, Long.parseLong(startedAt));
+    }
+  }
+
+  private String selectProperty(DbClient dbClient, DbSession dbSession, String key) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/platform/StartupMetadataProvider.java]: MAJOR Make "selectProperty" a "static" method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

File: server/sonar-server/src/main/java/org/sonar/server/platform/StartupMetadataProvider.java [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+    PropertyDto prop = dbClient.propertiesDao().selectGlobalProperty(dbSession, key);
+    checkState(prop != null, "Property %s is missing in database", key);
+    checkState(!isBlank(prop.getValue()), "Property %s is set but empty in database", key);
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/platform/StartupMetadataProvider.java]: MAJOR NullPointerException might be thrown as 'prop' is nullable here [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2259)

File: server/sonar-server/src/main/java/org/sonar/server/platform/StartupMetadataProvider.java [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+        cache = load(dbClient);
+      }
+    }
+    return cache;
+  }
+
+  /**
+   * Generate and persist in database
+   */
+  private StartupMetadata create(UuidFactory uuidFactory, System2 system, DbClient dbClient) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/platform/StartupMetadataProvider.java]: MAJOR Make "create" a "static" method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

File: server/sonar-server/src/main/java/org/sonar/server/platform/platformlevel/PlatformLevel.java [hunk 9]
```diff
@@ -123,6 +129,24 @@ protected void add(Object... objects) {
...
+  /**
+   * Add a component to container only if the server node is marked as "startupLeader" (cluster disabled
+   * or first node of the cluster to be started).
+   *
+   * @throws IllegalStateException if called from PlatformLevel1, when cluster settings are not loaded
+   */
+  protected void addIfStartupLeader(Object... objects) {
+    Optional<Cluster> cluster = getOptional(Cluster.class);
+    checkState(cluster.isPresent(), "Cluster settings not loaded yet");
+    if (cluster.get().isStartupLeader()) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/platform/platformlevel/PlatformLevel.java]: BLOCKER Call "cluster.isPresent()" before accessing the value. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS3655)

File: server/sonar-server/src/main/java/org/sonar/server/platform/cluster/ClusterProperties.java [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.server.platform.cluster;
+
+import com.google.common.collect.ImmutableList;
+import java.util.List;
+import org.sonar.api.PropertyType;
+import org.sonar.api.config.PropertyDefinition;
+
+public class ClusterProperties {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/platform/cluster/ClusterProperties.java]: MAJOR Add a private constructor to hide the implicit public one. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1118)

**Comment by SonarTech** [comment]
SonarQube analysis reported 24 issues
* BLOCKER 3 blocker
* CRITICAL 3 critical
* MAJOR 15 major
* MINOR 3 minor

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. BLOCKER [PlatformDatabaseMigration.java#L113](https://github.com/SonarSource/sonarqube/blob/dd62d7cd0198b68aa8eb3232d917a40f01cb91cb/server/sonar-server/src/main/java/org/sonar/server/platform/db/migrations/PlatformDatabaseMigration.java#L113): Catch Exception instead of Throwable. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1181)
1. BLOCKER [PlatformServletContextListener.java#L48](https://github.com/SonarSource/sonarqube/blob/dd62d7cd0198b68aa8eb3232d917a40f01cb91cb/server/sonar-server/src/main/java/org/sonar/server/platform/web/PlatformServletContextListener.java#L48): Catch Exception instead of Throwable. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1181)
1. CRITICAL [EmbeddedDatabase.java#L106](https://github.com/SonarSource/sonarqube/blob/dd62d7cd0198b68aa8eb3232d917a40f01cb91cb/server/sonar-server/src/main/java/org/sonar/server/platform/db/EmbeddedDatabase.java#L106): Remove this hard-coded password. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2068)
1. CRITICAL [MasterServletFilter.java#L44](https://github.com/SonarSource/sonarqube/blob/dd62d7cd0198b68aa8eb3232d917a40f01cb91cb/server/sonar-server/src/main/java/org/sonar/server/platform/web/MasterServletFilter.java#L44): Make this "public static INSTANCE" field final [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1444)
1. MAJOR [DatabaseMigrator.java#L52](https://github.com/SonarSource/sonarqube/blob/dd62d7cd0198b68aa8eb3232d917a40f01cb91cb/server/sonar-server/src/main/java/org/sonar/server/platform/db/migrations/DatabaseMigrator.java#L52): Remove this unused method parameter "unused". [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1172)
1. MAJOR [DatabaseMigrator.java#L107](https://github.com/SonarSource/sonarqube/blob/dd62d7cd0198b68aa8eb3232d917a40f01cb91cb/server/sonar-server/src/main/java/org/sonar/server/platform/db/migrations/DatabaseMigrator.java#L107): Use an "instanceof" comparison instead. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1872)
1. MAJOR [MasterServletFilter.java#L44](https://github.com/SonarSource/sonarqube/blob/dd62d7cd0198b68aa8eb3232d917a40f01cb91cb/server/sonar-server/src/main/java/org/sonar/server/platform/web/MasterServletFilter.java#L44): Make INSTANCE a static final constant or non-public and provide accessors if needed. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AClassVariableVisibilityCheck)
1. MAJOR [PlatformServletContextListener.java#L58](https://github.com/SonarSource/sonarqube/blob/dd62d7cd0198b68aa8eb3232d917a40f01cb91cb/server/sonar-server/src/main/java/org/sonar/server/platform/web/PlatformServletContextListener.java#L58): Make "stopQuietly" a "static" method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2325)
1. MAJOR [RailsAppsDeployer.java#L107](https://github.com/SonarSource/sonarqube/blob/dd62d7cd0198b68aa8eb3232d917a40f01cb91cb/server/sonar-server/src/main/java/org/sonar/server/platform/web/RailsAppsDeployer.java#L107): Make "prepareDir" a "static" method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2325)
1. MAJOR [ClassLoaderUtils.java#L108](https://github.com/SonarSource/sonarqube/blob/dd62d7cd0198b68aa8eb3232d917a40f01cb91cb/server/sonar-server/src/main/java/org/sonar/server/util/ClassLoaderUtils.java#L108): Refactor this code to not nest more than 3 if/for/while/switch/try statements. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS134)

File: server/sonar-server/src/main/java/org/sonar/server/util/ClassLoaderUtils.java [hunk 11]
```diff
@@ -110,7 +102,7 @@ public boolean apply(@Nullable String path) {
         Enumeration<JarEntry> entries = jar.entries();
         while (entries.hasMoreElements()) {
           String name = entries.nextElement().getName();
-          if (name.startsWith(rootDirectory) && predicate.apply(name)) {
+          if (name.startsWith(rootDirectory) && predicate.test(name)) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/util/ClassLoaderUtils.java]: MAJOR Refactor this code to not nest more than 3 if/for/while/switch/try statements. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS134)

File: server/sonar-server/src/main/java/org/sonar/server/platform/StartupMetadataProvider.java [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+      String startupId = selectProperty(dbClient, dbSession, CoreProperties.SERVER_ID);
+      String startedAt = selectProperty(dbClient, dbSession, PROPERTY_STARTED_AT);
+      return new StartupMetadata(startupId, Long.parseLong(startedAt));
+    }
+  }
+
+  private static String selectProperty(DbClient dbClient, DbSession dbSession, String key) {
+    PropertyDto prop = dbClient.propertiesDao().selectGlobalProperty(dbSession, key);
+    checkState(prop != null, "Property %s is missing in database", key);
+    checkState(!isBlank(prop.getValue()), "Property %s is set but empty in database", key);
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/platform/StartupMetadataProvider.java]: MAJOR NullPointerException might be thrown as 'prop' is nullable here [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2259)

**Comment by SonarTech** [comment]
SonarQube analysis reported 18 issues
* BLOCKER 3 blocker
* CRITICAL 3 critical
* MAJOR 9 major
* MINOR 3 minor

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. BLOCKER [PlatformDatabaseMigration.java#L113](https://github.com/SonarSource/sonarqube/blob/80b550e29bc8c71e82e4994bc6f616d8cfdd8bb3/server/sonar-server/src/main/java/org/sonar/server/platform/db/migrations/PlatformDatabaseMigration.java#L113): Catch Exception instead of Throwable. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1181)
1. BLOCKER [PlatformServletContextListener.java#L48](https://github.com/SonarSource/sonarqube/blob/80b550e29bc8c71e82e4994bc6f616d8cfdd8bb3/server/sonar-server/src/main/java/org/sonar/server/platform/web/PlatformServletContextListener.java#L48): Catch Exception instead of Throwable. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1181)
1. CRITICAL [EmbeddedDatabase.java#L106](https://github.com/SonarSource/sonarqube/blob/80b550e29bc8c71e82e4994bc6f616d8cfdd8bb3/server/sonar-server/src/main/java/org/sonar/server/platform/db/EmbeddedDatabase.java#L106): Remove this hard-coded password. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2068)
1. CRITICAL [MasterServletFilter.java#L44](https://github.com/SonarSource/sonarqube/blob/80b550e29bc8c71e82e4994bc6f616d8cfdd8bb3/server/sonar-server/src/main/java/org/sonar/server/platform/web/MasterServletFilter.java#L44): Make this "public static INSTANCE" field final [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1444)
1. MAJOR [DatabaseMigrator.java#L52](https://github.com/SonarSource/sonarqube/blob/80b550e29bc8c71e82e4994bc6f616d8cfdd8bb3/server/sonar-server/src/main/java/org/sonar/server/platform/db/migrations/DatabaseMigrator.java#L52): Remove this unused method parameter "unused". [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1172)
1. MAJOR [DatabaseMigrator.java#L107](https://github.com/SonarSource/sonarqube/blob/80b550e29bc8c71e82e4994bc6f616d8cfdd8bb3/server/sonar-server/src/main/java/org/sonar/server/platform/db/migrations/DatabaseMigrator.java#L107): Use an "instanceof" comparison instead. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1872)
1. MAJOR [MasterServletFilter.java#L44](https://github.com/SonarSource/sonarqube/blob/80b550e29bc8c71e82e4994bc6f616d8cfdd8bb3/server/sonar-server/src/main/java/org/sonar/server/platform/web/MasterServletFilter.java#L44): Make INSTANCE a static final constant or non-public and provide accessors if needed. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AClassVariableVisibilityCheck)
1. MAJOR [PlatformServletContextListener.java#L58](https://github.com/SonarSource/sonarqube/blob/80b550e29bc8c71e82e4994bc6f616d8cfdd8bb3/server/sonar-server/src/main/java/org/sonar/server/platform/web/PlatformServletContextListener.java#L58): Make "stopQuietly" a "static" method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2325)
1. MAJOR [RailsAppsDeployer.java#L107](https://github.com/SonarSource/sonarqube/blob/80b550e29bc8c71e82e4994bc6f616d8cfdd8bb3/server/sonar-server/src/main/java/org/sonar/server/platform/web/RailsAppsDeployer.java#L107): Make "prepareDir" a "static" method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2325)
1. MAJOR [PlatformDatabaseMigrationConcurrentAccessTest.java#L69](https://github.com/SonarSource/sonarqube/blob/80b550e29bc8c71e82e4994bc6f616d8cfdd8bb3/server/sonar-server/src/test/java/org/sonar/server/platform/db/migrations/PlatformDatabaseMigrationConcurrentAccessTest.java#L69): Remove this use of "Thread.sleep()". [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2925)

File: server/sonar-server/src/main/java/org/sonar/server/platform/ServerSettingsImpl.java [hunk 13]
```diff
@@ -35,17 +35,15 @@
  * <li>environment variables</li>
  * <li>system properties</li>
  * </ol>
- *
- * @since 2.12
  */
-public class WebServerSettings extends Settings implements ServerSettings {
+public class ServerSettingsImpl extends Settings implements ServerSettings {
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/platform/ServerSettingsImpl.java]: @simonbrandhof I don't understand this change

This class is actually specific to the WebServer (CE has it's own implementation called `ComputeEngineSettings`) and is not even in a package specific to the Web Server (such as `org.sonar.platform.web` or something)

File: server/sonar-server/src/main/java/org/sonar/server/platform/db/CheckDatabaseCharsetAtStartup.java [hunk 14]
```diff
@@ -17,7 +17,7 @@
  * along with this program; if not, write to the Free Software Foundation,
  * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
-package org.sonar.server.db;
+package org.sonar.server.platform.db;
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/platform/db/CheckDatabaseCharsetAtStartup.java]: @simonbrandhof what's the reasoning behind this change?

I don't see how the new package is better nor worse than the old one

**Pull request closed without merging**